### ISSUE TITLE
distinct() and group_by() do not use near() when making floating point comparisons

### ISSUE THREAD

**Issue by msto** [tidyverse/dplyr#5630]
Hi,

It appears that `distinct()` and `group_by()` do not use `near()` when making floating point comparisons. I think it would be more intuitive if these functions used `near()`, at least optionally, to permit grouping and de-duplicating of floats within machine epsilon (or other tolerance).

Could this be added? Or, is there a recommended way around this behavior? I apologize if this is intended behavior or documented elsewhere, I couldn't find mention of it in the issues here or the docs + stackoverflow.

Thanks!

```r
> df <- data.frame(x=c(sqrt(2) ^ 2, 2.0),
+                  y=c("a", "a"))
> df %>% distinct()
  x y
1 2 a
2 2 a
> df %>% group_by(x) %>% summarise(n=n())
`summarise()` ungrouping output (override with `.groups` argument)
# A tibble: 2 x 2
      x     n
  <dbl> <int>
1    2      1
2    2.     1
> df[1, "x"] == df[2, "x"]
[1] FALSE
> near(df[1, "x"], df[2, "x"])
[1] TRUE
> 
```

**Comment by msto** [comment]
Update -

1. I worked around this by rounding the columns in question. I was able to do this because it's a simple case where the values are all rational (specifically, an integer or an integer plus a multiple of 1/4, 1/3, or 1/2); it would likely not be the preferred solution for de-duplicating arbitrary floats.
2. I traced the behavior back to [`vctrs:::vctrs_unique_loc()`](https://github.com/r-lib/vctrs/blob/3110403857fa66ba46fdde6c2c5d9f05baabc050/src/dictionary.c#L181-L186), which uses a hash to distinguish unique values. I'm not sure what a good solution looks like here. Would it be out of scope for `dplyr` to implement an alternative uniqueness function for numeric vectors?
